discussment-core/src/main/java/org/danekja/discussment/core/service/imp/DefaultPostReputationService.java

```diff
@@ -36,41 +39,48 @@ public DefaultPostReputationService(UserPostReputationDao userPostReputationDao,
         this.accessControlService = accessControlService;
     }
 
-    public synchronized void addLike(Post post){
-        // todo: only one service must exist in the app
-        IDiscussionUser user = userService.getCurrentlyLoggedUser();
-        post.getPostReputation().addLike();
-        if (!userVotedOn(user, post)) {
-            userPostReputationDao.save(new UserPostReputation(user.getDiscussionUserId(), post, true));
+    public void addLike(Post post){
+        synchronized(monitor){
+            IDiscussionUser user = userService.getCurrentlyLoggedUser();
+            if (!userVotedOn(user, post)) {
+                post.getPostReputation().addLike();
+                userPostReputationDao.save(new UserPostReputation(user.getDiscussionUserId(), post, true));
+                postDao.save(post);
+            } else if (!userLiked(user, post)){
```

Comment:
This results into two unnecesarry calls to database in case user has previously voted (first userVotedOn, then userLiked).

Can be prevented by simply retrieving the vote from db. If it is null, user hasnt voted, if it isnt null, you can check how user voted before.